### PR TITLE
Support regex rewrite for path

### LINKED ISSUE #1191: [Feature] Support regex rewrite for route
### The New Feature

Describe the new feature you want to support clearly.

### Your scenes

Describe your use scenes (why need this feature).

### Your advice

Describe the advice or solution of this new feature.

### Environment

- MOSN Version

### PR DESCRIPTION
**Pull request by liangyuanpeng**
### Issues associated with this PR

closed #1191 
Your PR should present related issues you want to solve.

### Solutions
You should show your solutions about the issues in your PR, including the overall solutions, details and the changes. At this time, Chinese is allowed to describe these.

这个PR没有实现对正则表达式复杂度的计算`max_program_size`

在Envoy中是直接调用的`google/re2`这个库(cpp实现)  

而golang的`regexp`原生库是兼容`re2`的，所以直接用了regexp实现 (也没有发现go的re2包装实现)  

但是在golang中没有计算`max_program_size`的方法，我在golang/go的github仓库中提了一个issue,当前状态是打上了提案/建议的标签(没同意处理也没拒绝)  

关于`max_program_size`的内容我另外开了一个issue  #1201  

当前我能想到的只有用CGO实现.  或者等go官方支持? (可能要很久)

### UT result

```
Running tool: /usr/local/go/bin/go test -timeout 30s mosn.io/mosn/pkg/router -run ^(Test_RouteRuleImplBase_finalizePathHeader)$

ok  	mosn.io/mosn/pkg/router	0.021s
```

### Benchmark
If your code involves the processing of every request, you should give the Benchmark Result.

### Code Style
+ Make sure `Goimports` has run
+ Show `Golint` result

### CODE REVIEW

**Comment by liangyuanpeng** [comment]
@wangfakang  PTAL

File: pkg/router/base_rule.go [hunk 1]
```diff
@@ -180,15 +183,37 @@ func (rri *RouteRuleImplBase) matchRoute(headers api.HeaderMap, randomValue uint
...
+				headers.Set(protocol.MosnOriginalHeaderPathKey, path)
+				headers.Set(protocol.MosnHeaderPathKey, rri.prefixRewrite+path[len(matchedPath):])
+				log.DefaultLogger.Infof(RouterLogFormat, "routerule", "finalizePathHeader", "add prefix to path, prefix is "+rri.prefixRewrite)
+			}
+			return
+		}
+
+		// regex rewrite path if configured
+		if len(rri.regexRewrite.Pattern.Regex) > 1 {
+			reg := regexp.MustCompile(rri.regexRewrite.Pattern.Regex)
```
Comment (wangfakang) [inline pkg/router/base_rule.go:206]: if `rri.regexRewrite.Pattern.Regex`  cannot be parsed will caused `MustCompile` panic ?

Comment (liangyuanpeng) [reply pkg/router/base_rule.go:206]: i am sorry about that, maybe we should check regex in mosn just started. if got the error, panic mosn? 

like:  

```
reg,err  := regexp.Compile("\\  ? as")
if err!=nil{
   panic(err)
}
```

Comment (wangfakang) [reply pkg/router/base_rule.go:206]: Could  printe a error-level log  to alert the user.

File: pkg/router/base_rule.go [hunk 2]
```diff
@@ -200,15 +203,39 @@ func (rri *RouteRuleImplBase) matchRoute(headers api.HeaderMap, randomValue uint
...
+				headers.Set(protocol.MosnOriginalHeaderPathKey, path)
+				headers.Set(protocol.MosnHeaderPathKey, rri.prefixRewrite+path[len(matchedPath):])
+				log.DefaultLogger.Infof(RouterLogFormat, "routerule", "finalizePathHeader", "add prefix to path, prefix is "+rri.prefixRewrite)
+			}
+			return
+		}
+
+		// regex rewrite path if configured
+		if len(rri.regexRewrite.Pattern.Regex) > 1 {
+			reg, err := regexp.Compile(rri.regexRewrite.Pattern.Regex)
```
Comment (nejisama) [inline pkg/router/base_rule.go:226]: 每次都编译一次没有必要

Comment (liangyuanpeng) [reply pkg/router/base_rule.go:226]: Actually , in my test case . compile regex will be fast than cache regex to map.


Cache regex to map:

|count|cost time (ms)|  
|------|---|
|3,0000|299|
|30,0000|5233|
|60,0000|8952|

Just compile :

|count|cost time (ms)|  
|------|---|
|3,0000|139|
|30,0000|2011|
|60,0000|6755|

My test code:

```
func main() {
	text := `Hello World! 123 Go.`
	reg, err := regexp.Compile("(Hello)(.*)(Go)")
	if err != nil {
		panic(err)
	}

	regMap := make(map[string]*regexp.Regexp)
	regMap["service1"] = reg

	count := 300000

	var wg01 sync.WaitGroup

	wg01.Add(count)

	t01 := time.Now() //begin time

	for i := 0; i < count; i++ {
		go func() {
			regTmp, _:= regMap["service1"]
			regTmp.ReplaceAllString(text, "$3$2$1")
			wg01.Done()
		}()
	}
	wg01.Wait()
	t02 := time.Now()
	d01 := t02.Sub(t01)
	fmt.Printf("%d data ,cost time of map reg:%d ms \n", count, d01.Milliseconds())

	var wg02 sync.WaitGroup

	wg02.Add(count)

	t11 := time.Now() //begin time

	for i := 0; i < count; i++ {
		go func() {
			reg, err := regexp.Compile("(Hello)(.*)(Go)")
			if err != nil {
				panic(err)
			}
			reg.ReplaceAllString(text, "$3$2$1")
			wg02.Done()
		}()
	}
	wg02.Wait()
	t12 := time.Now()
	d11 := t12.Sub(t11)
	fmt.Printf("%d data ,cost time:%d ms \n", count, d11.Milliseconds())
}
```

Comment (nejisama) [reply pkg/router/base_rule.go:226]: rri.regexRewrite.Pattern.Regex is a static config, why you need a map to store the regex?

Comment (nejisama) [reply pkg/router/base_rule.go:226]: add a new field to store regex in rri ?_

Comment (liangyuanpeng) [reply pkg/router/base_rule.go:226]: Oh. you are right ! i will solved it. Thanks!

File: pkg/router/base_rule.go [hunk 3]
```diff
@@ -200,15 +214,35 @@ func (rri *RouteRuleImplBase) matchRoute(headers api.HeaderMap, randomValue uint
...
+				headers.Set(protocol.MosnOriginalHeaderPathKey, path)
+				headers.Set(protocol.MosnHeaderPathKey, rri.prefixRewrite+path[len(matchedPath):])
+				log.DefaultLogger.Infof(RouterLogFormat, "routerule", "finalizePathHeader", "add prefix to path, prefix is "+rri.prefixRewrite)
+			}
+			return
+		}
+
+		// regex rewrite path if configured
+		if len(rri.regexRewrite.Pattern.Regex) > 1 {
+			if rri.regexPattern.Match([]byte(path)) {
```
Comment (nejisama) [inline pkg/router/base_rule.go:237]: use regexp.MatchString instead of Match([]byte())

**Comment by nejisama** [comment]
unit test panic

```
=== RUN   Test_RouteRuleImplBase_finalizePathHeader/case1#01
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x18 pc=0x51be8f]
goroutine 27 [running]:
testing.tRunner.func1(0xc00012ad00)
	/usr/local/go/src/testing/testing.go:830 +0x392
panic(0xf07be0, 0x1bd5100)
	/usr/local/go/src/runtime/panic.go:522 +0x1b5
regexp.(*Regexp).doExecute(0x0, 0x0, 0x0, 0xc000287bc0, 0x13, 0x20, 0x0, 0x0, 0x0, 0x0, ...)
	/usr/local/go/src/regexp/exec.go:527 +0x6f
regexp.(*Regexp).doMatch(...)
	/usr/local/go/src/regexp/exec.go:514
regexp.(*Regexp).Match(...)
	/usr/local/go/src/regexp/regexp.go:474
mosn.io/mosn/pkg/router.(*RouteRuleImplBase).finalizePathHeader(0xc0002e8280, 0x11fa6a0, 0xc0002bd980, 0x108f4e5, 0x13)
	/go/src/mosn.io/mosn/pkg/router/base_rule.go:243 +0x47c
mosn.io/mosn/pkg/router.Test_RouteRuleImplBase_finalizePathHeader.func2(0xc00012ad00)
	/go/src/mosn.io/mosn/pkg/router/base_rule_test.go:342 +0x5a
testing.tRunner(0xc00012ad00, 0xc0002e00f0)
	/usr/local/go/src/testing/testing.go:865 +0xc0
created by testing.(*T).Run
	/usr/local/go/src/testing/testing.go:916 +0x35
```

File: pkg/router/base_rule.go [hunk 4]
```diff
@@ -200,15 +214,34 @@ func (rri *RouteRuleImplBase) matchRoute(headers api.HeaderMap, randomValue uint
...
+			if strings.HasPrefix(path, matchedPath) {
+				headers.Set(protocol.MosnOriginalHeaderPathKey, path)
+				headers.Set(protocol.MosnHeaderPathKey, rri.prefixRewrite+path[len(matchedPath):])
+				log.DefaultLogger.Infof(RouterLogFormat, "routerule", "finalizePathHeader", "add prefix to path, prefix is "+rri.prefixRewrite)
+			}
+			return
 		}
+
+		// regex rewrite path if configured
+		if len(rri.regexRewrite.Pattern.Regex) > 1 {
```
Comment (wangfakang) [inline pkg/router/base_rule.go:236]: Need check `rri.regexPattern`.

**Pull request merged**